Blockly Factory: Generate Category Xml

quachtina96 opened the pull request:
In order to allow users to add custom blocks to their toolbox and workspace configurations within Workspace Factory, we generate the xml for those categories.

- can generate category xml from Block Library (BlockExporterTools)
- can generate category xml from imported block definitions (Workspace Factory Generator)
- added helper functions for category xml generation to FactoryUtils
- can get Block Library category xml through BlockExporterController

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly/552)
<!-- Reviewable:end -->

quachtina96 commented:
@evd2014 @picklesrus @vicng

evd2014 commented:
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[demos/blocklyfactory/factory_utils.js, line 793 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfCxBpYdJ3OgPluK6:-KPJfCxBpYdJ3OgPluK7:b-r3yq4x) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L793)):*
> ```JavaScript
>  * block definitions.
>  *
>  * @param {!string} blockDefs - JavaScript block definition(s) each separated by a new
> ```

Now that your parsing is more robust you can update this comment (no longer have to be separated by new line).

---

*[demos/blocklyfactory/factory_utils.js, line 799 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJey5AVYV8BMVOWwHd:-KPJey5AVYV8BMVOWwHe:btwskne) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L799)):*
> ```JavaScript
> FactoryUtils.splitJSBlockDefs = function(blockDefs) {
>   var blockDefArray = [];
>   var blockDefs = goog.string.collapseBreakingSpaces(blockDefs);
> ```

Can you use a better name to distinguish blockDefs from blockDefArray? It's a little confusing when trying to tell the difference between the two. Maybe blockDefsFromFile? 

---

*[demos/blocklyfactory/factory_utils.js, line 802 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJeWuWama5-_sLL03d:-KPJeWuXj2FcVMQn1SUg:b-hq0nwx) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L802)):*
> ```JavaScript
>   var defStart = blockDefs.indexOf('Blockly.Blocks');
> 
>   while (blockDefs.indexOf('Blockly.Blocks', defStart) != -1) {
> ```

This while loop is a little confusing. It would be helpful if you could rewrite this loop so that if nextStart == -1, nextStart = blockDefs.length, and then continue with the rest of the loop as 1 case (the correct substring will be taken, pushed to the array, and then the loop will break because the while loop condition will no longer be true). 

---

*[demos/blocklyfactory/factory_utils.js, line 827 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfVuZc0sri0Ps4tuB:-KPJfVuZc0sri0Ps4tuC:bxjv43y) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L827)):*
> ```JavaScript
>  */
> FactoryUtils.splitJSONBlockDefs = function(blockDefs) {
>   var blockDefs = goog.string.collapseWhitespace(blockDefs);
> ```

See comment above for renaming blockDefs to distinguish between between blockDefs and blockDefArray.

---

*[demos/blocklyfactory/factory_utils.js, line 834 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJa5TMsgSEdAobQ-db:-KPJa5TNZe-TFD77Ssk1:bzc1j9f) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L834)):*
> ```JavaScript
>   // are balanced.
>   var i = 0;
>   while (i < blockDefs.length) {
> ```

Make this a for loop 
for (var i = 0; i < blockDefs.length; i++)

---

*[demos/blocklyfactory/factory_utils.js, line 839 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJb-ieXD8fKuRD3msL:-KPJb-ieXD8fKuRD3msM:b-h8oc5i) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L839)):*
> ```JavaScript
>       stack.push(currentChar);
>     }
>     else if (currentChar == '}') {
> ```

} else if (...) {

---

*[demos/blocklyfactory/factory_utils.js, line 840 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJaTQyAy1qpLfSjdqb:-KPJaTQz8tOBrO7w40CA:b-du2uzl) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L840)):*
> ```JavaScript
>     }
>     else if (currentChar == '}') {
>       stack.pop(currentChar);
> ```

Pop doesn't take any arguments. Stacks seems to just be counter for if the braces are balanced? If so, it would be best to keep track of it as an integer (++ instead of pushing, and -- instead of popping, and then you can check the value of the counter instead of the stack length)

---

*[demos/blocklyfactory/factory_utils.js, line 842 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJb43fd_dSQ-hQc54i:-KPJb43fd_dSQ-hQc54j:b-ruekpe) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L842)):*
> ```JavaScript
>       stack.pop(currentChar);
>     }
>     if (stack.length == 0 && i > 0 && currentChar != " ") {
> ```

It seems like this should be moved into the else if (currentChar == '}') because as your code currently stands, if there are any characters other than a space (including new lines and tabs, and other characters if the file isn't formatted correctly) between block definitions, then it's possible for the block definition plus some other text to be pushed to blockDefArray and then some more random text to be pushed to blockDefArray (because the braces are still technically balanced). 

This section of code should only be called the first time a set of braces are balanced, and that only happens when we find a closing brace, so we can move this inside of that block. Correct me if I'm wrong about this!

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 191 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfckus_TCuB3Jf-LL:-KPJfckus_TCuB3Jf-LM:bxjv43y) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L191)):*
> ```JavaScript
>  */
> FactoryGenerator.prototype.generateCategoryFromBlockDefs =
>     function(blockDefs, format, categoryName) {
> ```

See comment above for renaming blockDefs to distinguish between between blockDefs and blockDefArray.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 193 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfnG6CIKBxmvfbVAs:-KPJfnG6CIKBxmvfbVAt:bbpkvlo) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L193)):*
> ```JavaScript
>     function(blockDefs, format, categoryName) {
>   var blockTypes = [];
>   var blockDefArray;
> ```

Can blockDefArray be defined as a variable within each if/else if block? Even though there's no real scoping in Javascript, it would be more readable if the variable was declared only in the scope that it's used since it's never used outside of the if/else if blocks :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[demos/blocklyfactory/factory_utils.js, line 793 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfCxBpYdJ3OgPluK6:-KPJjjpPSLSxXpL8VWxd:b-896fix) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L793)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Now that your parsing is more robust you can update this comment (no longer have to be separated by new line).
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 799 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJey5AVYV8BMVOWwHd:-KPJkHVntHTAcC8j5l7C:b-896fix) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L799)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Can you use a better name to distinguish blockDefs from blockDefArray? It's a little confusing when trying to tell the difference between the two. Maybe blockDefsFromFile? 
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 802 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJeWuWama5-_sLL03d:-KPJzVlfnn-ctaW3oWBX:b-896fix) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L802)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> This while loop is a little confusing. It would be helpful if you could rewrite this loop so that if nextStart == -1, nextStart = blockDefs.length, and then continue with the rest of the loop as 1 case (the correct substring will be taken, pushed to the array, and then the loop will break because the while loop condition will no longer be true). 
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 827 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfVuZc0sri0Ps4tuB:-KPKRqJqDTONBjFdU6CZ:b-896fix) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L827)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> See comment above for renaming blockDefs to distinguish between between blockDefs and blockDefArray.
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 834 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJa5TMsgSEdAobQ-db:-KPKT3jtjTIHWcEJ1a0l:b-896fix) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L834)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Make this a for loop 
> for (var i = 0; i < blockDefs.length; i++)
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 839 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJb-ieXD8fKuRD3msL:-KPKT5vXdENpzY6VQEQ-:b-896fix) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L839)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> } else if (...) {
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 840 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJaTQyAy1qpLfSjdqb:-KPKT7qdMcKfYhk-ndu_:boq2z9o) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L840)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Pop doesn't take any arguments. Stacks seems to just be counter for if the braces are balanced? If so, it would be best to keep track of it as an integer (++ instead of pushing, and -- instead of popping, and then you can check the value of the counter instead of the stack length)
</details>
removed stack and replaced with counter

---

*[demos/blocklyfactory/factory_utils.js, line 842 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJb43fd_dSQ-hQc54i:-KPKTA0S-RCtswd_3iZp:b-vu32rg) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L842)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> It seems like this should be moved into the else if (currentChar == '}') because as your code currently stands, if there are any characters other than a space (including new lines and tabs, and other characters if the file isn't formatted correctly) between block definitions, then it's possible for the block definition plus some other text to be pushed to blockDefArray and then some more random text to be pushed to blockDefArray (because the braces are still technically balanced). 
> 
> This section of code should only be called the first time a set of braces are balanced, and that only happens when we find a closing brace, so we can move this inside of that block. Correct me if I'm wrong about this!
</details>
you're right! done!

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 191 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfckus_TCuB3Jf-LL:-KPJxONdNAkfsPnihGnO:b-896fix) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L191)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> See comment above for renaming blockDefs to distinguish between between blockDefs and blockDefArray.
</details>
Done.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 193 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfnG6CIKBxmvfbVAs:-KPJxQKHAfkAQJ9nfKpj:b-896fix) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L193)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Can blockDefArray be defined as a variable within each if/else if block? Even though there's no real scoping in Javascript, it would be more readable if the variable was declared only in the scope that it's used since it's never used outside of the if/else if blocks :)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552)*
<!-- Sent from Reviewable.io -->

vicng commented:
Review status: 0 of 4 files reviewed at latest revision, 19 unresolved discussions.

---

*[demos/blocklyfactory/factory_utils.js, line 827 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfVuZc0sri0Ps4tuB:-KPKcGB1l1fjb9e7128D:b-q0seqp) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L827)):*
<details><summary><i>Previously, quachtina96 (Tina Quach) wrote…</i></summary>
> Done.
</details>
Missing comment?

---

*[demos/blocklyfactory/factory_utils.js, line 742 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKXY8fOHztpYdVS6pD:-KPKXY8fOHztpYdVS6pE:buaqu3x) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L742)):*
> ```JavaScript
>   var indexOfStartBracket = blockDef.indexOf('[\'');
>   var indexOfEndBracket = blockDef.indexOf('\']');
>   return blockDef.substring(indexOfStartBracket + 2, indexOfEndBracket);
> ```

What happens if .substring is called with (-1, -1)?

---

*[demos/blocklyfactory/factory_utils.js, line 765 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKb6He5EmjsjMIs-xC:-KPKb6He5EmjsjMIs-xD:bfiqdqi) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L765)):*
> ```JavaScript
>  *    category.
>  * @param {string} categoryName - Name to use for the generated category.
>  * @param {!Blockly.Workspace} - Hidden blockly workspace.
> ```

Hidden workspace should not be passed in as a parameter.

This method should just take in "blocks" and "categoryName". 

---

*[demos/blocklyfactory/factory_utils.js, line 796 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKbfYjnIgyMlvtmPHB:-KPKbfYk4DxT_URoSt-u:b-obnftm) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L796)):*
> ```JavaScript
>  * @return {!Array.<string>} - Array of block definitions.
>  */
> FactoryUtils.splitJsBlockDefs = function(blockDefsString) {
> ```

Call this `parseJSBlockDefs`

---

*[demos/blocklyfactory/factory_utils.js, line 821 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKcBsGJhkJjP14t_cr:-KPKcBsGJhkJjP14t_cs:bmhc0x4) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L821)):*
> ```JavaScript
>  * @return {!Array.<string>} - Array of block definitions.
>  */
> FactoryUtils.splitJsonBlockDefs = function(blockDefsString) {
> ```

Rename to `parseJSONBlockDefs`

---

*[demos/blocklyfactory/factory_utils.js, line 829 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKcb_Ho2ROc6SAz6nG:-KPKcb_I9qnUMx9iN1Yj:b-ykaduj) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L829)):*
> ```JavaScript
>   // Iterate through the blockDefs string. Keep track of whether brackets
>   // are balanced.
>   for (var i = 0; i < blockDefsString.length; i++) {
> ```

What's the expected input? If blockDefsString is a JSON object, can't it just be parsed into a JS object and manipulated that way?

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 197 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKdZkKHBDsQVCC0lRz:-KPKdZkKHBDsQVCC0lS-:bw6r210) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L197)):*
> ```JavaScript
>   if (format == 'JSON') {
>     var blockDefArray = FactoryUtils.splitJsonBlockDefs(blockDefsString);
>     // Populate array of blocktypes and define each block.
> ```

Add line break above this line.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 201 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKda7ZmvQ78rMR0hDh:-KPKda7ZmvQ78rMR0hDi:b-pd442e) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L201)):*
> ```JavaScript
>       var json = JSON.parse(blockDef);
>       blockTypes.push(json.type);
>       FactoryUtils.defineBlockWithJson(json);
> ```

Is this the only place this method will be used? If so, just define that code in here.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 205 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKdYQKKJNHaZqKEsfk:-KPKdYQKKJNHaZqKEsfl:bw6r210) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L205)):*
> ```JavaScript
>   } else if (format == 'JavaScript') {
>     var blockDefArray = FactoryUtils.splitJsBlockDefs(blockDefsString);
>     // Populate array of block types.
> ```

Add line break above this line.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 210 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKdTwb01Nca4cwbH5W:-KPKdTwcHCDXkP20zFl5:bw6r210) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L210)):*
> ```JavaScript
>       blockTypes.push(blockType);
>     }
>     // Define all blocks.
> ```

Add line break above this line.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 214 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKeJ6ZJsZ_VusDEq6Q:-KPKeJ6_xC29j8aThquh:b-6dli5l) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L214)):*
> ```JavaScript
>   }
> 
>   return FactoryUtils.generateCategoryXml(blockTypes, categoryName,
> ```

This method is misleading since it mostly defines blocks based on format (which isn't in the method name or method comment) and then only this one line of code actually generates the category.

Also, `hiddenWorkspace` is passed in here, which doesn't really make sense to me as a reader why that's happening.

Try refactoring the code to be more readable. One suggestion would be to change the method name and its responsibilities.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[demos/blocklyfactory/factory_utils.js, line 827 \[r1\]](https://reviewable.io:443/reviews/google/blockly/552#-KPJfVuZc0sri0Ps4tuB:-KPKp_lLUBcfAixDfmQ7:bcgoaku) ([raw file](https://github.com/google/blockly/blob/54286855c403b58a35ace7b946b62ae371982e1a/demos/blocklyfactory/factory_utils.js#L827)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Missing comment?
</details>
was orginally going to comment what collapseWhitespace does, but actually I removed the calls to collapsing the whitespace now that I no longer need them.

---

*[demos/blocklyfactory/factory_utils.js, line 742 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKXY8fOHztpYdVS6pD:-KPKrmpDmtrZaLf8viBU:b-x8se3m) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L742)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> What happens if .substring is called with (-1, -1)?
</details>
Returns an empty string. I see what you're saying. When this was private, I made assumptions about what inputs this function would get. I've now added checks to see if the indexOf calls return -1, and throw and error in the case that either of the calls to indexOf returns -1.

---

*[demos/blocklyfactory/factory_utils.js, line 765 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKb6He5EmjsjMIs-xC:-KPKndSNTlXbvxDawS7t:b-gy9bju) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L765)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Hidden workspace should not be passed in as a parameter.
> 
> This method should just take in "blocks" and "categoryName". 
</details>
factory utils doesn't have its own hidden workspace, but both exporter tools and wfactory generator do. I talked to Katelyn about possibly giving the factoryutils its own hidden workspace and whether that would mean factoryutils should be an object rather than just a namespace. I'd prefer if factoryutils was just a set of functions for simplicity.

For now, I've made it so you pass in any workspace since it still needs a Blockly Workspace in order to get the defined block and its block's xml. I also added a comment but am not sure that the comment is too tied to the implementation

---

*[demos/blocklyfactory/factory_utils.js, line 796 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKbfYjnIgyMlvtmPHB:-KPKoNlDqDDFuM0dwxwE:b-896fix) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L796)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Call this `parseJSBlockDefs`
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 821 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKcBsGJhkJjP14t_cr:-KPKoO2AKJvbuLZzOMyK:b-896fix) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L821)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Rename to `parseJSONBlockDefs`
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 829 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKcb_Ho2ROc6SAz6nG:-KPKpf_8hCURvBJ9vrOy:b-jhbcmn) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L829)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> What's the expected input? If blockDefsString is a JSON object, can't it just be parsed into a JS object and manipulated that way?
</details>
blockDefsString is multiple JSON object string representations concatenated.
The native JSON.parse() function expects the whole string to be valid JSON.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 197 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKdZkKHBDsQVCC0lRz:-KPKsY-0V-tZYnUb41jH:b-896fix) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L197)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Add line break above this line.
</details>
Done.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 201 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKda7ZmvQ78rMR0hDh:-KPKsosezS5XzUxUyioE:bcbttu4) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L201)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Is this the only place this method will be used? If so, just define that code in here.
</details>
I've deleted the defineBlockWithJson method and decided it was small enough to include directly in the function.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 205 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKdYQKKJNHaZqKEsfk:-KPKszeGqNZEBSTe7OIE:b-896fix) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L205)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Add line break above this line.
</details>
Done.

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 210 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKdTwb01Nca4cwbH5W:-KPKszSWj6Tb-cC28l0T:b-kdv6am) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L210)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Add line break above this line.
</details>
Done. Will remember this for next time!

---

*[demos/blocklyfactory/workspacefactory/wfactory_generator.js, line 214 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKeJ6ZJsZ_VusDEq6Q:-KPKx5P1Y1BSvT9idRNt:bweypaf) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/workspacefactory/wfactory_generator.js#L214)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> This method is misleading since it mostly defines blocks based on format (which isn't in the method name or method comment) and then only this one line of code actually generates the category.
> 
> Also, `hiddenWorkspace` is passed in here, which doesn't really make sense to me as a reader why that's happening.
> 
> Try refactoring the code to be more readable. One suggestion would be to change the method name and its responsibilities.
</details>
I've refactored the code so that getCategoryFromBlockDefs is no longer in this  wfactory_generator and and i've made a new function in FactoryUtils for defining and getting the blocktypes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Also updated BlockExporterTools to use FactoryUtils.getCategoryXml in generating toolbox from block lib

---

Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552#-:-KPLAmPN3lu5iO3Co2u_:bilufrq)*
<!-- Sent from Reviewable.io -->

evd2014 commented:
Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions.

---

*[demos/blocklyfactory/block_exporter_controller.js, line 95 \[r3\]](https://reviewable.io:443/reviews/google/blockly/552#-KPOXX0ZPH3cI7Vb6gXz:-KPOXX0ZPH3cI7Vb6gY-:b-ayxfly) ([raw file](https://github.com/google/blockly/blob/999c64dea883467d3cd2d8289a5b8dcf6f1688b7/demos/blocklyfactory/block_exporter_controller.js#L95)):*
> ```JavaScript
> /**
>  * Unselect a block for export.
>  * @private
> ```

Either remove @private tag or add _ to the end of the function name.

---

*[demos/blocklyfactory/block_exporter_tools.js, line 197 \[r3\]](https://reviewable.io:443/reviews/google/blockly/552#-KPOVyJ1pqx3o5a26GWw:-KPOVyJ2Xkr9w3fL9svo:b-rrmdzt) ([raw file](https://github.com/google/blockly/blob/999c64dea883467d3cd2d8289a5b8dcf6f1688b7/demos/blocklyfactory/block_exporter_tools.js#L197)):*
> ```JavaScript
>   }
> 
>   // If there are no blocks in library, append dummy category.
> ```

Your code currently appends the dummy category even if there are blocks in the library, so this comment is misleading. Either update the comment to show that it always appends the dummy category, or add a quick check to see if there are no blocks in the XML library (blockXmlMap.length == 0)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions.

---

*[demos/blocklyfactory/block_exporter_controller.js, line 95 \[r3\]](https://reviewable.io:443/reviews/google/blockly/552#-KPOXX0ZPH3cI7Vb6gXz:-KPOovc2EdMd79GlgjiZ:b-nhtdca) ([raw file](https://github.com/google/blockly/blob/999c64dea883467d3cd2d8289a5b8dcf6f1688b7/demos/blocklyfactory/block_exporter_controller.js#L95)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Either remove @private tag or add _ to the end of the function name.
</details>
deleted the entire function--don't need it after all

---

*[demos/blocklyfactory/block_exporter_tools.js, line 197 \[r3\]](https://reviewable.io:443/reviews/google/blockly/552#-KPOVyJ1pqx3o5a26GWw:-KPOqM4HNLtfeOkmXKI1:b-896fix) ([raw file](https://github.com/google/blockly/blob/999c64dea883467d3cd2d8289a5b8dcf6f1688b7/demos/blocklyfactory/block_exporter_tools.js#L197)):*
<details><summary><i>Previously, evd2014 (Emma Dauterman) wrote…</i></summary>
> Your code currently appends the dummy category even if there are blocks in the library, so this comment is misleading. Either update the comment to show that it always appends the dummy category, or add a quick check to see if there are no blocks in the XML library (blockXmlMap.length == 0)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552)*
<!-- Sent from Reviewable.io -->

vicng commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> after minor changes

---

Review status: 0 of 4 files reviewed at latest revision, 14 unresolved discussions.

---

*[demos/blocklyfactory/block_exporter_tools.js, line 197 \[r3\]](https://reviewable.io:443/reviews/google/blockly/552#-KPOVyJ1pqx3o5a26GWw:-KPPDkRf0fYtJ2VKwUwH:b-wx1i1j) ([raw file](https://github.com/google/blockly/blob/999c64dea883467d3cd2d8289a5b8dcf6f1688b7/demos/blocklyfactory/block_exporter_tools.js#L197)):*
<details><summary><i>Previously, quachtina96 wrote…</i></summary>
> Done.
</details>
Missing commit?

---

*[demos/blocklyfactory/factory_utils.js, line 829 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKcb_Ho2ROc6SAz6nG:-KPPAqmeIWS9E1r9r69N:biw3azh) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L829)):*
<details><summary><i>Previously, quachtina96 wrote…</i></summary>
> blockDefsString is multiple JSON object string representations concatenated.
> The native JSON.parse() function expects the whole string to be valid JSON.
</details>
Update the method comment to explain what the expected input is and maybe give an example.

---

*[demos/blocklyfactory/factory_utils.js, line 759 \[r3\]](https://reviewable.io:443/reviews/google/blockly/552#-KPPAHskz-9twOu2oo_h:-KPPAHskz-9twOu2oo_i:beodpkz) ([raw file](https://github.com/google/blockly/blob/999c64dea883467d3cd2d8289a5b8dcf6f1688b7/demos/blocklyfactory/factory_utils.js#L759)):*
> ```JavaScript
>  */
> FactoryUtils.generateCategoryXml =
>     function(blocks, categoryName) {
> ```

Nit: I think this can be moved back to the previous line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552#-:-KPPDgMTrmt0MBJt9017:b-861mez)*
<!-- Sent from Reviewable.io -->

quachtina96 commented:
Review status: 0 of 4 files reviewed at latest revision, 14 unresolved discussions.

---

*[demos/blocklyfactory/block_exporter_tools.js, line 197 \[r3\]](https://reviewable.io:443/reviews/google/blockly/552#-KPOVyJ1pqx3o5a26GWw:-KPPOoT4EpjHwGYnW6G1:bbw9nay) ([raw file](https://github.com/google/blockly/blob/999c64dea883467d3cd2d8289a5b8dcf6f1688b7/demos/blocklyfactory/block_exporter_tools.js#L197)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Missing commit?
</details>
The desired changes are there--I accidentally published before pushing

---

*[demos/blocklyfactory/factory_utils.js, line 829 \[r2\]](https://reviewable.io:443/reviews/google/blockly/552#-KPKcb_Ho2ROc6SAz6nG:-KPPQQ3j-vHKHoWJtcsK:b-896fix) ([raw file](https://github.com/google/blockly/blob/cf39a3efa9d44c1e7976c85080caf37819848a1c/demos/blocklyfactory/factory_utils.js#L829)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Update the method comment to explain what the expected input is and maybe give an example.
</details>
Done.

---

*[demos/blocklyfactory/factory_utils.js, line 759 \[r3\]](https://reviewable.io:443/reviews/google/blockly/552#-KPPAHskz-9twOu2oo_h:-KPPOmkAcFpzdOr5HfSB:b-896fix) ([raw file](https://github.com/google/blockly/blob/999c64dea883467d3cd2d8289a5b8dcf6f1688b7/demos/blocklyfactory/factory_utils.js#L759)):*
<details><summary><i>Previously, vicng wrote…</i></summary>
> Nit: I think this can be moved back to the previous line.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552)*
<!-- Sent from Reviewable.io -->

evd2014 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/552#-:-KPPdaBVGEzTBlNBZSew:bnfp4nl)*
<!-- Sent from Reviewable.io -->
